### PR TITLE
perf(codegen): deopt exact callbacks and captured updates

### PR DESCRIPTION
**Pull request by proggeramlug**
## Summary

- select a deliberately narrow family of exact additive arrow callbacks and compile private callback bodies with cold-arm deoptimization
- admit unused-result prefix/postfix `++` and `--` on compiler-proven captured boxes; plain numbers step inline while TDZ and ToNumeric coercion poison the fast loop before observable fallback
- cache admitted ECS array handles in the loop preheader and remove per-iteration fingerprint reloads and loop polls only while the exact hot path cannot collect
- fail closed through runtime identity and capture-layout validation, ordinary guarded-loop fallback, exact next-index resume, and an EH-scope exclusion

## Performance

Apple M1 Mac mini, Node 26.5.1, unchanged codehz/ecs workload:

- accumulation confirmation from the exact additive callback change: Perry 0.082606 ms, Node 0.086533 ms, 0.9542x, 11/11 wins
- read-only screen after admitting the benchmark `count++` callback: Perry 0.188263 ms, Node 0.052334 ms, 3.597x, 0/3 wins

The read-only result is recorded as an intermediate mechanism result, not a parity claim. Later stacked PRs remove the remaining object/closure overhead.

## Validation

- `cargo fmt --all -- --check`
- `cargo test --release -p perry-codegen`: 1,238 library tests passed, 1 ignored; every integration binary passed
- `cargo test --release -p perry --test versioned_indexed_loop_callback_deopt`: passed, including normal and forced-evacuation execution plus emitted-LLVM assertions
- unchanged ECS semantic/checksum oracles passed in every measured process
- rebased conflict-free onto current `main` (`d3d260295`)

This PR was updated but not merged by the performance investigation.

## Summary by CodeRabbit

* **Performance Improvements**
  * Reduced per-iteration checks for eligible additive callbacks in versioned indexed loops.
  * Improved garbage-collection handling during optimized callback execution.

* **Bug Fixes**
  * Preserved safe fallback behavior for unsupported callbacks and edge cases.
  * Ensured optimized loops resume at the correct index when falling back.
  * Maintained consistent results across garbage-collection modes and mixed value types.

* **Tests**
  * Added coverage for callback specialization, direct dispatch, fallback scenarios, exceptions, and GC behavior.

### CODE REVIEW

**Comment by proggeramlug** [comment]
Review follow-up is in a6c610f57: the runtime build, fixture compile, and both child-process arms remove the established collector environment overrides before the forced arm is applied. The release integration passed its normal and forced-moving-GC executions on the Mac mini.

File: crates/perry/tests/versioned_indexed_loop_callback_deopt.rs [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+    "PERRY_GC_SCAVENGE_NURSERY_MB",
+    "PERRY_GC_MOVING_SAFEPOINT",
+    "PERRY_GC_MOVING_LOOP_POLLS",
+    "PERRY_GC_FORCE_EVACUATE",
+    "PERRY_GC_VERIFY_EVACUATION",
+    "PERRY_CONSERVATIVE_STACK_SCAN",
+    "PERRY_WRITE_BARRIERS",
+    "PERRY_GC_INCREMENTAL",
+    "PERRY_GC_HEAP_LIMIT",
+];
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply crates/perry/tests/versioned_indexed_loop_callback_deopt.rs:25]: Verified and intentionally not added. PERRY_GEN_GC_EVACUATE was deleted in #7611; crates/perry-runtime/src/gc/mod.rs explicitly documents the deletion at the former gate, and no live runtime code reads the variable. The current forced-moving control is PERRY_GC_FORCE_EVACUATE, which is normalized and then set by the forced arm. Adding a dead variable to the canonical override list would preserve a stale configuration contract without changing the test. The release integration passed both arms on the Mac mini after normalization.

**Comment by proggeramlug** [comment]
Landed on `main` via #8808.
